turbot/resource_turbot_file.go

```diff
@@ -0,0 +1,253 @@
+package turbot
+
+import (
+	"fmt"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/terraform-providers/terraform-provider-turbot/apiClient"
+	"github.com/terraform-providers/terraform-provider-turbot/helpers"
+)
+
+var fileProperties = []interface{}{"parent", "tags", "akas"}
+var metadataProperties = []interface{}{"title", "description"}
+
+func getFileUpdateProperties() []interface{} {
+	excludedProperties := []string{"type"}
+	return helpers.RemoveProperties(resourceProperties, excludedProperties)
+}
+
+func resourceTurbotFile() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceTurbotFileCreate,
+		Read:   resourceTurbotFileRead,
+		Update: resourceTurbotFileUpdate,
+		Delete: resourceTurbotFileDelete,
+		Exists: resourceTurbotFileExists,
+		Importer: &schema.ResourceImporter{
+			State: resourceTurbotFileImport,
+		},
+		Schema: map[string]*schema.Schema{
+			// aka of the parent resource
+			"parent": {
+				Type:     schema.TypeString,
+				Required: true,
+				// when doing a diff, the state file will contain the id of the parent but the config contains the aka,
+				// so we need custom diff code
+				DiffSuppressFunc: suppressIfAkaMatches("parent_akas"),
+			},
+			// when doing a read, fetch the parent akas to use in suppressIfAkaMatches
+			"parent_akas": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"title": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"data": {
+				Type:             schema.TypeString,
+				Required:         true,
+				DiffSuppressFunc: suppressIfDataMatches,
+			},
+			"metadata": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				DiffSuppressFunc: suppressIfDataMatches,
+			},
+			"tags": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"akas": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+func resourceTurbotFileExists(d *schema.ResourceData, meta interface{}) (b bool, e error) {
+	client := meta.(*apiClient.Client)
+	id := d.Id()
+	return client.ResourceExists(id)
+}
+
+func resourceTurbotFileCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*apiClient.Client)
+	description := d.Get("description")
+	var err error
+	// build input map to pass to mutation
+	input, err := buildFileInput(d, fileProperties)
+	// set type property
+	input["type"] = "tmod:@turbot/turbot#/resource/types/file"
```

Comment:
why doesn't buildFileInput do this?

Reply:
ah because we do not need for update?

Reply:
Yes because then it will be pass on for Update operation, which is not valid